src/datasource.ts

```diff
@@ -404,34 +428,62 @@ export default class IrondbDatasource {
       });
   }
 
-  getRollupSpan(options, start, end, isCaql) {
-    // Pick a reasonable period for CAQL
-    // We assume will use something close the request interval
-    // unless it would produce more than maxDataPoints / 8
-    // CAQL analytics at one point per pixel is almost never what
-    // someone will want.
-    let estdp = Math.floor(((end - start) * 1000) / options.intervalMs);
-    if (estdp > options.maxDataPoints / 2) {
-      estdp = options.maxDataPoints / 2;
-    }
-    let period = Math.ceil((end - start) / estdp);
-    // The period is in the right realm now, force align to something
-    // that will make it pretty.
-    const align = [86400, 3600, 1800, 1200, 900, 300, 60, 30, 15, 10, 5, 1];
-    for (const j in align) {
-      if (period > align[j]) {
-        period = Math.floor(period / align[j]) * align[j];
-        break;
+  static readonly MAX_DATAPOINTS_THRESHOLD = 1.5;
+  static readonly MIN_DURATION_MS_FETCH = 1;
+  static readonly MIN_DURATION_MS_CAQL = 60 * 1000;
+  static readonly ROLLUP_ALIGN_MS = _.map([1, 60, 3600, 86400], x => x * 1000);
+  static readonly ROLLUP_ALIGN_MS_1DAY = 86400 * 1000;
+
+  static checkRollupAligned(rollupMs) {
+    for (const alignMs of IrondbDatasource.ROLLUP_ALIGN_MS) {
+      if (rollupMs < alignMs) {
+        if (alignMs % rollupMs !== 0) {
+          throw new Error('Unaligned rollup period requested');
+        }
       }
     }
-    if (isCaql) {
-      if (period < 60) {
-        period = 60;
-      } else {
-        period = period - (period % 60);
+    if (rollupMs > IrondbDatasource.ROLLUP_ALIGN_MS_1DAY && rollupMs % IrondbDatasource.ROLLUP_ALIGN_MS_1DAY !== 0) {
+      throw new Error('Unaligned rollup period requested');
+    }
+  }
+
+  getRollupSpan(options, start, end, isCaql, leafData) {
+    log(() => `getRollupSpan() intervalMs = ${options.intervalMs}, maxDataPoints = ${options.maxDataPoints}`);
+    log(() => `getRollupSpan() ${isCaql ? 'CAQL' : '/fetch'} ${JSON.stringify(leafData)}`);
+    let rolluptype = leafData.rolluptype;
+    const metricrollup = leafData.metricrollup;
+    if (rolluptype !== 'automatic' && _.isEmpty(metricrollup)) {
+      rolluptype = 'automatic';
+      log(() => `getRollupSpan() defaulting to automatic`);
+    }
+    let rollupMs;
+    if (rolluptype === 'exact') {
+      const exactMs = parseDurationMs(metricrollup);
+      const exactDatapoints = Math.floor(((end - start) * 1000) / exactMs);
+      log(() => `getRollupSpan() exactMs = ${exactMs}, exactDatapoints = ${exactDatapoints}`);
+      if (exactDatapoints * IrondbDatasource.MAX_DATAPOINTS_THRESHOLD > options.maxDataPoints) {
```

Comment:
Let's make MAX_DATAPOINTS_THRESHOLD and MAX_EXACT_DATAPOINTS_THRESHOLD two separate constants that both happen to be 1.5 right now.  I have a feeling people are going to say: "I said 5ms rollups, stop telling me I can't" so we might need to up this threshold here, but there's no reason to up the other case.

Reply:
I think you multiplied on the wrong side of this. this should be:
`exactDatapoints > options.maxDataPoints * IrondbDatasource.MAX_EXACT_DATAPOINTS_THRESHOLD`